src/server/frontend_wayland/xdg_activation_v1.cpp

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "xdg_activation_v1.h"
+#include "wl_surface.h"
+#include "wl_seat.h"
+#include "wl_client.h"
+#include "mir/main_loop.h"
+#include "mir/scene/surface.h"
+#include "mir/shell/shell.h"
+#include "mir/shell/focus_controller.h"
+#include "mir/wayland/protocol_error.h"
+#include "mir/log.h"
+#include <random>
+#include <chrono>
+#include <uuid.h>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace msh = mir::shell;
+
+namespace
+{
+/// Time in milliseconds that the compositor will wait before invalidating a token
+auto constexpr timeout_ms = std::chrono::milliseconds(3000);
+
+std::string generate_token()
+{
+    uuid_t uuid;
+    uuid_generate(uuid);
+    return { reinterpret_cast<char*>(uuid), 4 };
+}
+}
+
+namespace mir
+{
+namespace frontend
+{
+class XdgActivationV1 : public wayland::XdgActivationV1::Global
+{
+public:
+    XdgActivationV1(
+        struct wl_display* display,
+        std::shared_ptr<msh::Shell> const& shell,
+        std::shared_ptr<msh::FocusController> const& focus_controller,
+        std::shared_ptr<MainLoop> const& main_loop);
+
+    std::string const& create_token();
+
+    bool try_consume_token(std::string const& token);
+
+private:
+    class Instance : public wayland::XdgActivationV1
+    {
+    public:
+        Instance(
+            struct wl_resource* resource,
+            mf::XdgActivationV1* xdg_activation_v1,
+            std::shared_ptr<msh::Shell> const& shell,
+            std::shared_ptr<msh::FocusController> const& focus_controller,
+            std::shared_ptr<MainLoop> const& main_loop);
+
+    private:
+        void get_activation_token(struct wl_resource* id) override;
+
+        void activate(std::string const& token, struct wl_resource* surface) override;
+
+        mf::XdgActivationV1* xdg_activation_v1;
+        std::shared_ptr<msh::Shell> shell;
+        std::shared_ptr<msh::FocusController> const focus_controller;
+        std::shared_ptr<MainLoop> main_loop;
+    };
+
+    struct Token
+    {
+        std::string token;
+        std::unique_ptr<time::Alarm> alarm;
+    };
+
+    void bind(wl_resource* resource) override;
+
+    std::shared_ptr<msh::Shell> shell;
+    std::shared_ptr<msh::FocusController> const focus_controller;
+    std::shared_ptr<MainLoop> main_loop;
+    std::vector<Token> pending_tokens;
+    std::mutex mutex;
```

Comment:
Perhaps things could be clearer:
```suggestion
    // pending_tokens is accessed by an alarm callback, not only the wayland thread
    std::mutex pending_tokens_mutex;
    std::vector<Token> pending_tokens;
```